### PR TITLE
many: clarify gadget role-usage consistency checks for UC16/18 vs UC20

### PR DESCRIPTION
**Pull request by pedronis**
some of the commits are preparation cleanups for the changes themselves, for example to avoid exposing changing internal details to tests or to increase test coverage.

this also has some start of supporting properly multi-volume situations, roles are considered/counted across volumes but enforced to come from the same at the end.

this also stops calling the model Model parameters constraints, using "model characteristics" instead.

### CODE REVIEW

**Review by mvo5 — COMMENTED** [review]
Thanks, this looks good but one question about the TODO/XXX you added there.

**Review by bboozzoo — APPROVED** [review]
Thanks.

File: gadget/validate.go [hunk 1]
```diff
@@ -77,74 +70,104 @@ func validateEncryptionSupport(info *Info) error {
...
+		SystemData: nil,
+		SystemSave: nil,
+	}
+
+	xvols := ""
+	if len(vols) != 1 {
+		xvols = " across volumes"
+	}
+
+	// TODO: is this too strict for old gadgets?
```
[the remark this answers is not in the document]
Comment (pedronis) [reply gadget/validate.go:91]: As we discussed the plan is to advertize the changes on the forum after this gets into edge

File: gadget/validate.go [hunk 2]
```diff
@@ -77,73 +70,104 @@ func validateEncryptionSupport(info *Info) error {
 	return nil
 }
 
+type roleInst struct {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply gadget/validate.go:73]: it stands for instance, instance of the role, does that make sense?

[the remark this answers is not in the document]
Comment (bboozzoo) [reply gadget/validate.go:73]: Yeah, I guess we can leave it as it is then.

[the remark this answers is not in the document]
Comment (pedronis) [reply gadget/validate.go:73]: it's not used a lot, I can use the fuller name roleInstance

File: gadget/validate.go [hunk 3]
```diff
@@ -159,98 +182,88 @@ func validateReservedLabels(vs *VolumeStructure) error {
...
+		// make sure that all roles come from the same volume
+		// TODO:UC20: there is more to do in order to support multi-volume situations
+
+		// if SystemSeed is unset we must have failed earlier
+		seedVolName := roles[SystemSeed].volName
 
+		for _, otherRole := range []string{SystemBoot, SystemData, SystemSave} {
+			ri := roles[otherRole]
+			if ri != nil && ri.volName != seedVolName {
+				return fmt.Errorf("system-boot|data|save are expected to share the same volume as system-seed")
```
Comment (anonymouse64) [inline gadget/validate.go:234]: minor preference to have a longer error message here, but not a big deal either way

```suggestion
				return fmt.Errorf("system-boot, system-data, and system-|save are expected to share the same volume as system-seed")
```

File: gadget/validate_test.go [hunk 4]
```diff
@@ -43,65 +43,121 @@ func (s *validateGadgetTestSuite) SetUpTest(c *C) {
...
+		}
+	}
+
+	gi, err := gadget.InfoFromGadgetYaml([]byte(h), nil)
+	c.Assert(err, IsNil)
+	return gi
+}
+
+func (s *validateGadgetTestSuite) TestVolumeRulesConsistencyNoModel(c *C) {
+	ginfo := func(hasSeed bool, label string) *gadget.Info {
```
Comment (anonymouse64) [inline gadget/validate_test.go:133]:
```suggestion
	ginfo := func(hasSeed bool, dataLabel string) *gadget.Info {
```

(this makes it more clear that the label here is the label for the data/writable partition)

File: gadget/validate_test.go [hunk 5]
```diff
@@ -43,65 +43,121 @@ func (s *validateGadgetTestSuite) SetUpTest(c *C) {
...
+		seed := "-"
+		if hasSeed {
+			seed = ""
+		}
+		return rolesYaml(c, label, seed, "-")
```
Comment (anonymouse64) [inline gadget/validate_test.go:138]:
```suggestion
		return rolesYaml(c, dataLabel, seed, "-")
```

File: gadget/validate_test.go [hunk 6]
```diff
@@ -298,18 +353,91 @@ volumes:
...
 	}
-	ginfo, err := gadget.ReadInfo(s.dir, constraints)
+
+	ginfo, err := gadget.ReadInfo(s.dir, mod)
+	c.Assert(err, IsNil)
+	err = gadget.Validate(ginfo, mod, nil)
+	c.Assert(err, ErrorMatches, "model requires system-seed partition, but no system-seed or system-data partition found")
+}
+
+func (s *validateGadgetTestSuite) TestValidateSystemRoleSplitAcrossVolumes(c *C) {
```
Comment (anonymouse64) [inline gadget/validate_test.go:366]: it's fine if you don't want to add it to this pr, but I think we should add a test case with all of the ubuntu-* partitions/roles on one volume and another volume with unrelated partitions on it to demonstrate that multiple volumes works with uc20, as it's my understanding that this pr would enable that setup

**Review by anonymouse64 — APPROVED** [review]
small little nitpicks, but looks good!

File: gadget/validate_test.go (hunk 6 above)
Comment (pedronis) [reply gadget/validate_test.go:366]: yea, this seems follow up material

**Comment by pedronis** [comment]
these are the required tests that failed in the last spread run:  one set is store issues, the other seems to be recurrent flakiness(?)

- google:ubuntu-18.04-64:tests/main/security-dev-input-event-denied
- google:ubuntu-20.10-64:tests/main/apt-hooks
- google:ubuntu-20.10-64:tests/main/snap-advise-command
- google:ubuntu-core-18-64:tests/main/security-dev-input-event-denied

**Pull request merged**